### PR TITLE
Do not deaggregate MIR

### PR DESCRIPTION
**Pull request by cjgillot**
This turns out to simplify a lot of things.
I haven't checked the consequences for miri yet.

cc @JakobDegen 
r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
I reopened #35259 because there are conflicting desires

* @eddyb would prefer to go all the way and remove `Aggregate` from the MIR entirely
    * this avoids having two ways to do the same thing, but needs a marker for when a value is fully constructed
* @JakobDegen would prefer to do this PR and never deaggregate
    * this is easier on analyses and some optimizations, but now we have two ways to do the same thing (assigning each field separately vs assigning them all at once)

The discussion that caused this PR: https://rust-lang.zulipchat.com/#narrow/stream/136281-t-lang.2Fwg-unsafe-code-guidelines/topic/Changing.20enum.20variant.20by.20writing.20to.20field

**Comment by JakobDegen** [comment]
One thing we can do is keep deaggregating structs, but not enums. Then there won't be two ways of doing the same thing, as long as we don't allow `SetDiscriminant` for enums at all

**Comment by oli-obk** [comment]
So... basically the reason deaggregation was enabled for everything but arrays was to work towards named return value optimization. Instead of building all the components and then doing a single aggregate assignment at the end, you can assign to sub-places immediately and even end up with multiple layers of function calls that assign to a place in the outermost call.

I guess we can still do that as part of nrvo when it decides a return value should get that opt run on it.

**Comment by JakobDegen** [comment]
> So... basically the reason deaggregation was enabled for everything but arrays was to work towards named return value optimization. Instead of building all the components and then doing a single aggregate assignment at the end, you can assign to sub-places immediately and even end up with multiple layers of function calls that assign to a place in the outermost call.

Honestly, this is a problem we should fix via SSA (or something similar, not via deaggregation)

**Comment by oli-obk** [comment]
Makes sense. So I guess the only thing left is to give codegen backends the ability to deaggregate so they don't have to implement both aggregate rvalues, field assignments and setdiscriminant unless they want to

File: compiler/rustc_const_eval/src/interpret/step.rs [hunk 1]
```diff
@@ -199,12 +200,48 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
 
             Aggregate(box ref kind, ref operands) => {
-                assert!(matches!(kind, mir::AggregateKind::Array(..)));
-
-                for (field_index, operand) in operands.iter().enumerate() {
-                    let op = self.eval_operand(operand, None)?;
-                    let field_dest = self.place_field(&dest, field_index)?;
-                    self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                self.write_uninit(&dest)?;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/step.rs:203]: This seems worth factoring into a separate function, e.g. `write_aggregate`, probably in `place.rs`.

File: compiler/rustc_const_eval/src/interpret/step.rs [hunk 2]
```diff
@@ -199,12 +200,48 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                match kind {
+                    mir::AggregateKind::Struct { .. } => {
+                        for (field_index, operand) in operands.iter().enumerate() {
+                            let field_dest = self.place_field(&dest, field_index)?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                    }
+                    mir::AggregateKind::Union { active_field_index, .. } => {
+                        debug_assert_eq!(operands.len(), 1);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/step.rs:213]: Can we have similar debug assertions in the other arms, ensuring that we have the expected number of operands?

File: compiler/rustc_const_eval/src/interpret/step.rs [hunk 3]
```diff
@@ -199,12 +200,48 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                    mir::AggregateKind::Generator(..) => {
+                        for (field_index, operand) in operands.iter().enumerate() {
+                            let field_dest = self.place_field(&dest, field_index)?;
+                            let op = self.eval_operand(operand, Some(field_dest.layout))?;
+                            self.copy_op(&op, &field_dest, /*allow_transmute*/ false)?;
+                        }
+                        self.write_discriminant(VariantIdx::from_u32(0), &dest)?;
+                    }
+                    _ => {
+                        for (field_index, operand) in operands.iter().enumerate() {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/step.rs:239]: Basically each match arm has this same for loop. Can't that common code be factored out somehow?

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 4]
```diff
@@ -1196,17 +1196,26 @@ pub enum AggregateKind<'tcx> {
...
+    },
+    Union {
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        active_field_index: Field,
+        user_ty: Option<UserTypeAnnotationIndex>,
+    },
+
+    Closure(DefId, SubstsRef<'tcx>),
+    Generator(DefId, SubstsRef<'tcx>, hir::Movability),
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/syntax.rs:1218]: Please make this split a separate PR, it seems like something we should discuss and review on its own

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: Is this split even a good idea? At least in Miri, I think some of the code duplication is caused by the redundancy in the Struct/Enum/Union cases. Generally the ADT handling in rustc is treating structs as "1-variant enums" and that tends to avoid code duplication. Why diverge from that pattern here?

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: I don't know the motivation for it, but whatever it is, it seems orthogonal to changing deaggregation.

Comment (Aurel300) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: We would appreciate this change in Prusti, but I don't know how useful/annoying this is for other consumers of the MIR.

Comment (cjgillot) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: I did this split in a separate commit to have this discussion.

The three cases are deaggegated/interpreted/codegened in very different ways, so keeping them merged required to disentangle them manually by checking `adt_def(def_id).is_enum()`.

For instance, deaggregation considered the struct and union field accesses to be direct, without the implicit downcast to variant 0.

If we stop deaggregating, this split is probably not that useful: the interpreter and codegen could perform the implicit downcast to variant 0 in all cases, and it should be no-op for struct and enums.

@Aurel300 could you elaborate on the patterns that this enables/inhibits?

Comment (Aurel300) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: In our encoding we also treat struct and enums differently (and we don't yet support unions), so it's common to see this in one `match` arm:

https://github.com/viperproject/prusti-dev/blob/17b071c41c1b502d739ccf414a2445a1bad8a3f6/prusti-viper/src/encoder/snapshot/encoder.rs#L715

And in another:

https://github.com/viperproject/prusti-dev/blob/17b071c41c1b502d739ccf414a2445a1bad8a3f6/prusti-viper/src/encoder/snapshot/encoder.rs#L744

Comment (cjgillot) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: @Aurel300 your examples show usage of `TyKind::Adt`. This discussion is about `mir::AggregateKind::Adt`. Am I missing something?

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: > The three cases are deaggegated/interpreted/codegened in very different ways, 

Are they? At least for struct and enum, we iterate over all the fields of one variant of an ADT, and assign them. That's pretty similar.

Comment (Aurel300) [reply compiler/rustc_middle/src/mir/syntax.rs:1218]: @cjgillot Ah, disregard my comment: I incorrectly assumed the change was on `TyKind`. For encoding a `AggregateKind` we still need some special logic *only if* the ADT is an enum, but then perform common actions in both enum and struct cases.

**Comment by nnethercote** [comment]
The wins here outweigh the losses, even if you ignore the handful of very good results among secondary benchmarks that skew the overall results.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_codegen_ssa/src/mir/rvalue.rs [hunk 5]
```diff
@@ -106,31 +107,31 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
             mir::Rvalue::Aggregate(ref kind, ref operands) => {
-                let (dest, active_field_index) = match **kind {
-                    mir::AggregateKind::Adt(adt_did, variant_index, _, _, active_field_index) => {
-                        dest.codegen_set_discr(bx, variant_index);
-                        if bx.tcx().adt_def(adt_did).is_enum() {
-                            (dest.project_downcast(bx, variant_index), active_field_index)
-                        } else {
-                            (dest, active_field_index)
-                        }
+                let (variant_index, variant_dest, active_field_index) = match **kind {
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/mir/rvalue.rs:110]: Cg_clif needs to be updated too. Will do this over at the cg_clif repo.

**Comment by Manishearth** [comment]
This caused a regression: https://github.com/rust-lang/rust/issues/107678

Should this be reverted?

**Comment by saethlin** [comment]
The fix is https://github.com/rust-lang/rust/pull/107688, I think this PR just makes the root cause easier to access. I've seen similar-looking ICEs in other crates (associated type vs concrete type) but was never able to minimize them.
